File: android/src/main/kotlin/com/eopeter/fluttermapboxnavigation/boundings/map/application/Codec.kt
package com.eopeter.fluttermapboxnavigation.boundings.map.application

import com.eopeter.fluttermapboxnavigation.core.*
import io.flutter.plugin.common.StandardMessageCodec
import java.io.ByteArrayOutputStream
import java.nio.ByteBuffer

@Suppress("UNCHECKED_CAST")
object MapApiCodec : StandardMessageCodec() {
    override fun readValueOfType(type: Byte, buffer: ByteBuffer): Any? {
        return when (type) {
            128.toByte() -> {
                return (readValue(buffer) as? List<Any?>)?.let {
                    QueriedFeature.fromList(it)
                }
            }
            129.toByte() -> {
                return (readValue(buffer) as? List<Any?>)?.let {
                    RenderedQueryGeometry.fromList(it)
                }
            }
            130.toByte() -> {
                return (readValue(buffer) as? List<Any?>)?.let {
                    RenderedQueryOptions.fromList(it)
                }
            }
            131.toByte() -> {
                return (readValue(buffer) as? List<Any?>)?.let {
                    ScreenCoordinate.fromList(it)
                }
            }
            else -> super.readValueOfType(type, buffer)
        }
    }
    override fun writeValue(stream: ByteArrayOutputStream, value: Any?) {
        when (value) {
            is QueriedFeature -> {
                stream.write(128)
                writeValue(stream, value.toList())
            }
            is RenderedQueryGeometry -> {
                stream.write(129)
                writeValue(stream, value.toList())
            }
            is RenderedQueryOptions -> {
                stream.write(130)
                writeValue(stream, value.toList())
            }
            is ScreenCoordinate -> {
                stream.write(131)
                writeValue(stream, value.toList())
            }
            else -> super.writeValue(stream, value)
        }
    }
}

File: lib/src/embedded/boundings/map/application/codec.dart
part of mapbox_navigation_flutter;

class MapAPICodec extends StandardMessageCodec {
  const MapAPICodec();
  @override
  void writeValue(WriteBuffer buffer, Object? value) {
    if (value is QueriedFeature) {
      buffer.putUint8(128);
      writeValue(buffer, value.encode());
    } else if (value is RenderedQueryGeometry) {
      buffer.putUint8(129);
      writeValue(buffer, value.encode());
    } else if (value is RenderedQueryOptions) {
      buffer.putUint8(130);
      writeValue(buffer, value.encode());
    } else if (value is ScreenCoordinate) {
      buffer.putUint8(131);
      writeValue(buffer, value.encode());
    } else {
      super.writeValue(buffer, value);
    }
  }

  @override
  Object? readValueOfType(int type, ReadBuffer buffer) {
    switch (type) {
      case 128:
        return QueriedFeature.decode(readValue(buffer)!);
      case 129:
        return RenderedQueryGeometry.decode(readValue(buffer)!);
      case 130:
        return RenderedQueryOptions.decode(readValue(buffer)!);
      case 131:
        return ScreenCoordinate.decode(readValue(buffer)!);
      default:
        return super.readValueOfType(type, buffer);
    }
  }
}
